### PR TITLE
Compute snow metrics, all tiles, with filtering and filling implemented

### PR DESCRIPTION
**Pull request by charparr**
Welp, this is a giant PR!

This PR essentially "does the work" meaning it successfully downloads data, preprocesses the data into a datacube, does some filtering and filling, and then computes snow metrics for all tiles that are required to map the full Alaska domain.

To review this PR, please start with the README. It is in decent shape. But, here is the quick start:

- Clone this repo on GINA's Elephant machine, then make the conda env per the README instructions.
- Activate, set some env vars

```sh
conda activate viirs-snow
export SNOW_YEAR=2015
export INPUT_DIR=/export/datadir/"$USER"_viirs_snow/VIIRS_L3_snow_cover
export OUTPUT_DIR=/export/datadir/"$USER"_viirs_snow/VIIRS_snow_metrics
export SCRATCH_DIR=/export/datadir/"$USER"_viirs_snow/scratch
export MALLOC_TRIM_THRESHOLD_=0
```

Now, to run this pipeline execute the runner script. For example this command:
```sh
python snow_metric_runner.py --preprocess --filter_fill --compute_masks --compute_metrics --postprocess
```
will execute the parts of the pipeline passed as arguments to the script. In this example, the download option is omitted.

If you'd like to test the download component, just add `--download` to this chain. Downloading all the data will take a long time (perhaps 24 hours) while NSIDC prepares the data orders. If you'd like to skip this step you can use my `INPUT_DIR` like this:
```sh
export INPUT_DIR=/export/datadir/cparr4_viirs_snow/VIIRS_L3_snow_cover
```
But if you **do** want to download data go ahead and pick a different snow year! Something between 2013 and 2022.

Each script that executes will write log files to your `$HOME` directory.

If/when the script completes (expect ~24 hours), you'll find the "pot of gold" data in `/export/datadir/"$USER"_viirs_snow/scratch/"$SNOW_YEAR"/reprojected_merged_single_metric_geotiffs` and so to examine these results I suggest copying them to a local machine for QGIS-ing, or just read and inspect via your favorite raster tooling like GDAL / Jupyter Notebooks. Note that the $OUTPUT_DIR is placeholder for an archival location - we aren't ready for that yet.

When examining the results please perform some high-level QC:
 - Do the data open as expected?
 - Are the value ranges plausible? Note that some metrics are a count of days (leap years are ON), while others indicate a specific "day of the snow year" which is this zany DOY that ranges between 213 and 500-something.
 - Do the geographic patterns make sense: snowier at high elevations / latitudes?
 
I'd also ask you to read through the code as well. I've tried to document + make things legible...but if any of this seems obtuse please comment! 

There is a lot here, so I'll leave it at that and please just ask questions.

### CODE REVIEW

**Comment by jdelamere** [comment]
2015 finished processing on Elephant with the single example Charlie provided. Great so far.

**FULL 2015 PROCESSING (no data download, used Charlie's data)**
- cparr and delamere output present almost identically for file size. Files are almost identical when differencing but not quite as there are "lines" across the state where the pixel differences go from 0 to some small integer (in units of days) (this is really just a few pixels across many).
- Readme has 11 files, output has 10 (Number of Cloud Days is missing).
- reprojected_merged_uncertainty_geotiffs is empty. Does one run the gather_uncertainty_data.py to get this? (Was also empty in cparr's directory)
- In subdirectories leading to the merged product, many files have 2 underscores with nothing in between (doesn't matter, just unusual to see __)(h06v03__first_snow_day_2015.tif). 
- postprocess.log file has an ERROR posted 
_ERROR:root:
INFO:root:Creating output file that is 4636P x 3890L._
- Is timing information available on how long it takes to do each step, other than file timestamps?
- Should we consider having the log files post to somewhere other than home?

**JUST DATA DOWNLOAD**
- I've started that (also via screen) and expect this to take a long time. Log says:
_INFO:root:order ID: 5000005671107
INFO:root:status URL: https://n5eil02u.ecs.nsidc.org/egi/request/5000005671107
INFO:root:Status code from order response URL is 201
INFO:root:Data request 1 is submitting...
INFO:root:Initial request status is processing
INFO:root:Status is not complete. Trying again._

README.md
- Should we provide the earthdata URL for quick access?

**Review by kyleredilla — COMMENTED** [review]
I processed the 2015 snow year with exceptional ease, using the README and cparr’s data mentioned in the PR. Code looks and drives like a dream. I did a high level review of the code and found it very tidy and easy to follow - only found some minor suggestions to make.

I have been playing around with the 10 merged / reprojected geotiffs in QGIS, and things look good to my eye. One thing that surprised me though is seeing CSS ranges of ~20-30 days in the interior, along with start days well after 365 - is that expected? I would have thought the longest continuous snow season would be closer to ~100 d, but still less than no_snow_days, and with earlier start days, etc. But I am pretty sure I am just lacking the proper understanding of CSS. 

This looks really close to being ready for shipment! All of Jen’s comments are things I agree with / wonder about, too. Nice work Charlie 👍

File: compute_dark_and_cloud_metrics.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+
+    Args:
+        snow_is_on_at_dusk (xr.DataArray): boolean indicating if snow is on at dusk.
+        snow_is_on_at_dawn (xr.DataArray): boolean indicating if snow is on at dawn.
+    Returns:
+        xr.DataArray: integer array indicating the snow transition case during the obscured period. 1 = no change, 2 = snow flipped on, 3 = snow flipped off.
+    """
+    # no change in snow condition during the obscured period
+    # return true when conditions are identical before and after obscured period
+    snow_did_not_flip = ~(snow_is_on_at_dusk ^ snow_is_on_at_dawn)
```
[the remark this answers is not in the document]
Comment (charparr) [reply compute_dark_and_cloud_metrics.py:198]: So weird `^` feels like a carryover from a different era of computing or something

File: compute_dark_and_cloud_metrics.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+def get_last_obscured_occurence_index(dark_on):
+    """Get the last occurrence of the obscured condition.
+
+    Returns the time index of the last chronological occurrence of the obscured condition. Ranges between 0 and 365.
+    Args:
+        dark_on (xr.DataArray): Obscured grid cells.
+    Returns:
+        xr.DataArray: The last occurrence of the obscured condition."""
+    dark_on_reverse_time = dark_on.isel(time=slice(None, None, -1))
```
[the remark this answers is not in the document]
Comment (charparr) [reply compute_dark_and_cloud_metrics.py:73]: Me either (until I learned that ☝️) and it came in handy! Just another python built-in that I had zero idea about.

File: compute_dark_and_cloud_metrics.py
[the remark this answers is not in the document]
Comment (charparr) [reply compute_dark_and_cloud_metrics.py:1]: Thanks Kyle! Lots of functions, gotta keep 'em straight.

File: download.py [hunk 3]
```diff
@@ -82,6 +87,7 @@ def determine_tiles_for_bbox(bounding_box):
 
     tile_strs = []
     for tile in tile_list:
+        # hacky zero padding
```
[the remark this answers is not in the document]
Comment (charparr) [reply download.py:90]: ticketed

File: download.py [hunk 4]
```diff
@@ -261,35 +282,35 @@ def construct_request(
         tuple: containing a list of API endpoints and a dictionary of download parameters.
     """
     base_url = "https://n5eil02u.ecs.nsidc.org/egi/request"
-
+    # this is the sauce where we ask NSIDC to do additional upstream processing
```
[the remark this answers is not in the document]
Comment (charparr) [reply download.py:285]: Yeah, white picket fence kinda thing :)

**Comment by charparr** [comment]
@jdelamere OK I've resolved the little stuff and ticketed everything else! Go ahead and hit squash and merge when you feel ready and please let me know if you want assistance!

**Pull request merged**